File: src/cli/commands/exclude.ts
import type { Command } from "commander";
import { Registry } from "../../core/registry.js";
import { loadConfig, saveConfig } from "../../utils/config.js";
import { log } from "../../utils/logger.js";
import { resolveProjectSlug } from "../resolve.js";

export function registerExcludeCommand(program: Command): void {
	const exclude = program.command("exclude").description("Manage exclude patterns for a project");

	exclude
		.command("add <pattern...>")
		.description("Add exclude patterns (e.g. grafana-data)")
		.option("--in <slug>", "Project slug")
		.option("--global", "Apply to all projects")
		.action((patterns: string[], opts: { in?: string; global?: boolean }) => {
			const normalized = patterns.map((p) => (p.endsWith("/**") ? p : `${p}/**`));

			if (opts.global) {
				const config = loadConfig();
				const current = config.globalExclude ?? [];
				const added: string[] = [];
				for (const p of normalized) {
					if (!current.includes(p)) {
						current.push(p);
						added.push(p);
					}
				}
				if (added.length === 0) {
					log.dim("All patterns already in global excludes.");
					return;
				}
				config.globalExclude = current;
				saveConfig(config);
				log.success(`Added global excludes: ${added.join(", ")}`);
				log.dim("Run: graphmind clean --all && graphmind build --all --full");
				return;
			}

			const resolved = resolveProjectSlug(opts.in);
			if (!resolved) {
				log.error("Not in a registered project. Use --in <slug> or --global.");
				process.exitCode = 1;
				return;
			}

			const registry = new Registry();
			const project = registry.get(resolved);
			if (!project) {
				log.error(`Project "${resolved}" not found.`);
				process.exitCode = 1;
				return;
			}

			const current = project.exclude ?? [];
			const added: string[] = [];
			for (const p of normalized) {
				if (!current.includes(p)) {
					current.push(p);
					added.push(p);
				}
			}

			if (added.length === 0) {
				log.dim("All patterns already excluded.");
				return;
			}

			registry.updateProject(resolved, { exclude: current });
			log.success(`Added to "${resolved}": ${added.join(", ")}`);
			log.dim("Run: graphmind clean && graphmind build --full");
		});

	exclude
		.command("remove <pattern...>")
		.description("Remove exclude patterns")
		.option("--in <slug>", "Project slug")
		.option("--global", "Remove from global excludes")
		.action((patterns: string[], opts: { in?: string; global?: boolean }) => {
			const toRemove = new Set(patterns.flatMap((p) => [p, p.endsWith("/**") ? p : `${p}/**`]));

			if (opts.global) {
				const config = loadConfig();
				const current = config.globalExclude ?? [];
				const filtered = current.filter((e) => !toRemove.has(e));
				const removed = current.length - filtered.length;
				if (removed === 0) {
					log.dim("No matching global patterns found.");
					return;
				}
				config.globalExclude = filtered;
				saveConfig(config);
				log.success(`Removed ${removed} global pattern(s).`);
				return;
			}

			const resolved = resolveProjectSlug(opts.in);
			if (!resolved) {
				log.error("Not in a registered project. Use --in <slug> or --global.");
				process.exitCode = 1;
				return;
			}

			const registry = new Registry();
			const project = registry.get(resolved);
			if (!project) {
				log.error(`Project "${resolved}" not found.`);
				process.exitCode = 1;
				return;
			}

			const current = project.exclude ?? [];
			const filtered = current.filter((e) => !toRemove.has(e));
			const removed = current.length - filtered.length;

			if (removed === 0) {
				log.dim("No matching patterns found.");
				return;
			}

			registry.updateProject(resolved, { exclude: filtered });
			log.success(`Removed ${removed} pattern(s) from "${resolved}".`);
		});

	exclude
		.command("list")
		.description("Show current exclude patterns")
		.option("--in <slug>", "Project slug")
		.action((opts: { in?: string }) => {
			const config = loadConfig();
			const globalExclude = config.globalExclude ?? [];

			if (globalExclude.length > 0) {
				console.log("\n  Global excludes:");
				for (const e of globalExclude) {
					console.log(`    ${e}`);
				}
			}

			const resolved = resolveProjectSlug(opts.in);
			if (resolved) {
				const registry = new Registry();
				const project = registry.get(resolved);
				if (project) {
					const excludes = project.exclude ?? [];
					if (excludes.length > 0) {
						console.log(`\n  Project "${resolved}" excludes:`);
						for (const e of excludes) {
							console.log(`    ${e}`);
						}
					}
				}
			}

			if (globalExclude.length === 0 && !resolved) {
				log.dim("No exclude patterns configured.");
			}
			console.log();
		});
}

File: crates/graphmind-cli/src/commands/exclude.rs
use graphmind_config::{load_config, save_config, Registry, resolve_project_slug};
use colored::Colorize;

pub fn add(patterns: &[String], in_slug: Option<&str>, global: bool) {
    if global {
        let mut config = load_config();
        for p in patterns {
            if !config.global_exclude.contains(p) {
                config.global_exclude.push(p.clone());
            }
        }
        save_config(&config);
        println!(
            "{} Added {} global exclude pattern(s)",
            "OK".green().bold(),
            patterns.len()
        );
    } else {
        let slug = match resolve_project_slug(&[in_slug]) {
            Some(s) => s,
            None => {
                eprintln!(
                    "{} No project specified and none could be resolved",
                    "Error:".red().bold()
                );
                std::process::exit(1);
            }
        };

        Registry::update_project(&slug, |p| {
            for pattern in patterns {
                if !p.exclude.contains(pattern) {
                    p.exclude.push(pattern.clone());
                }
            }
        });
        println!(
            "{} Added {} exclude pattern(s) to {}",
            "OK".green().bold(),
            patterns.len(),
            slug.cyan()
        );
    }
}

pub fn remove(patterns: &[String], in_slug: Option<&str>, global: bool) {
    if global {
        let mut config = load_config();
        config.global_exclude.retain(|e| !patterns.contains(e));
        save_config(&config);
        println!(
            "{} Removed {} global exclude pattern(s)",
            "OK".green().bold(),
            patterns.len()
        );
    } else {
        let slug = match resolve_project_slug(&[in_slug]) {
            Some(s) => s,
            None => {
                eprintln!(
                    "{} No project specified and none could be resolved",
                    "Error:".red().bold()
                );
                std::process::exit(1);
            }
        };

        Registry::update_project(&slug, |p| {
            p.exclude.retain(|e| !patterns.contains(e));
        });
        println!(
            "{} Removed {} exclude pattern(s) from {}",
            "OK".green().bold(),
            patterns.len(),
            slug.cyan()
        );
    }
}

pub fn list(in_slug: Option<&str>, global: bool) {
    let config = load_config();

    if global {
        if config.global_exclude.is_empty() {
            println!("{}", "No global exclude patterns.".dimmed());
        } else {
            println!("{} Global exclude patterns:", ">>".cyan().bold());
            for p in &config.global_exclude {
                println!("  {p}");
            }
        }
        return;
    }

    let slug = match resolve_project_slug(&[in_slug]) {
        Some(s) => s,
        None => {
            eprintln!(
                "{} No project specified and none could be resolved",
                "Error:".red().bold()
            );
            std::process::exit(1);
        }
    };

    let project = match Registry::get(&slug) {
        Some(p) => p,
        None => {
            eprintln!("{} Project {} not found", "Error:".red().bold(), slug);
            std::process::exit(1);
        }
    };

    if project.exclude.is_empty() {
        println!("{} No exclude patterns for {}", "!".yellow(), slug);
    } else {
        println!("{} Exclude patterns for {}:", ">>".cyan().bold(), slug.cyan());
        for p in &project.exclude {
            println!("  {p}");
        }
    }

    if !config.global_exclude.is_empty() {
        println!("\n{} Global exclude patterns (also applied):", ">>".cyan().bold());
        for p in &config.global_exclude {
            println!("  {p}");
        }
    }
}
